### PR TITLE
Perf/lexer faster slow get char and size

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

**Comment by cor3ntin** [comment]
Do you have benchmarks?
I wonder if we should use a small struct instead of pair, it would be slightly easier to maintain I think.
Looks good otherwise.

**Comment by tbaederr** [comment]
> Do you have benchmarks? I wonder if we should use a small struct instead of pair, it would be slightly easier to maintain I think. Looks good otherwise.

See the commit: https://github.com/llvm/llvm-project/pull/70543/commits/c9d34dae319de3eed1a23c23ff7b6d7673a04b79

**Comment by serge-sans-paille** [comment]
@cor3ntin yep, looks better with a small struct, updated pushed

File: clang/lib/Lex/Lexer.cpp [hunk 1]
```diff
@@ -1964,11 +1969,14 @@ bool Lexer::LexIdentifierContinue(Token &Result, const char *CurPtr) {
...
 bool Lexer::isHexaLiteral(const char *Start, const LangOptions &LangOpts) {
-  unsigned Size;
-  char C1 = Lexer::getCharAndSizeNoWarn(Start, Size, LangOpts);
+  auto CharAndSize1 = Lexer::getCharAndSizeNoWarn(Start, LangOpts);
+  char C1 = CharAndSize1.Char;
   if (C1 != '0')
     return false;
-  char C2 = Lexer::getCharAndSizeNoWarn(Start + Size, Size, LangOpts);
+
+  auto CharAndSize2 =
```
Comment (cor3ntin) [inline clang/lib/Lex/Lexer.cpp:1977]: You can reuse `CharAndSize` here

File: clang/lib/Lex/Lexer.cpp [hunk 2]
```diff
@@ -2012,15 +2020,15 @@ bool Lexer::LexNumericConstant(Token &Result, const char *CurPtr) {
 
   // If we have a digit separator, continue.
   if (C == '\'' && (LangOpts.CPlusPlus14 || LangOpts.C23)) {
-    unsigned NextSize;
-    char Next = getCharAndSizeNoWarn(CurPtr + Size, NextSize, LangOpts);
+    auto NextCharAndSize = getCharAndSizeNoWarn(CurPtr + Size, LangOpts);
```
Comment (cor3ntin) [inline clang/lib/Lex/Lexer.cpp:2023]: Structured binding here?

File: clang/lib/Lex/Lexer.cpp [hunk 3]
```diff
@@ -2085,8 +2093,9 @@ const char *Lexer::LexUDSuffix(Token &Result, const char *CurPtr,
       unsigned Consumed = Size;
       unsigned Chars = 1;
       while (true) {
-        unsigned NextSize;
-        char Next = getCharAndSizeNoWarn(CurPtr + Consumed, NextSize, LangOpts);
+        auto NextCharAndSize =
```
Comment (cor3ntin) [inline clang/lib/Lex/Lexer.cpp:2096]: Structured binding here?

File: clang/lib/Lex/DependencyDirectivesScanner.cpp [hunk 4]
```diff
@@ -565,10 +565,9 @@ Scanner::cleanStringIfNeeded(const dependency_directives_scan::Token &Tok) {
   const char *BufPtr = Input.begin() + Tok.Offset;
   const char *AfterIdent = Input.begin() + Tok.getEnd();
   while (BufPtr < AfterIdent) {
-    unsigned Size;
-    Spelling[SpellingLength++] =
-        Lexer::getCharAndSizeNoWarn(BufPtr, Size, LangOpts);
-    BufPtr += Size;
+    auto CharAndSize = Lexer::getCharAndSizeNoWarn(BufPtr, LangOpts);
```
Comment (cor3ntin) [inline clang/lib/Lex/DependencyDirectivesScanner.cpp:568]: Structured binding here?

File: clang/lib/Lex/Lexer.cpp (hunk 1 above)
Comment (serge-sans-paille) [reply clang/lib/Lex/Lexer.cpp:1977]: Yeah, but I find it less confusing like this.

Comment (cor3ntin) [reply clang/lib/Lex/Lexer.cpp:1977]: I won't insist!

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks

**Comment by rorth** [comment]
This patch broke builders like [Solaris/amd64](https://lab.llvm.org/staging/#/builders/8/builds/761) that include `clang-tools-extra`.

**Comment by nico** [comment]
This breaks building: http://45.33.8.238/linux/121901/step_4.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by nico** [comment]
Reverted in 1c876ff5155c4feeb2b2885eb3e6abda17c4b7f4 for now.

**Comment by serge-sans-paille** [comment]
recommited as 8116b6dce70ef284f9a0895b0f9f45bfa9a3ade7

**Pull request merged**